spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -1495,6 +1501,10 @@ private void fixTxOffsetsIfNeeded() {
 			}
 			else {
 				records = pollConsumer();
+				if (this.pendingRecordsAfterError != null && !isPaused()) {
```

Comment:
I might be completely wrong about this - lots of moving parts here. But let's say that, after an error is handled and the `pendingRecordsAfterError` property is set with the remaining records, the container is set to `pause`. If I'm not mistaken, this is how it would go:

* the consumer is paused
* the next poll returns empty and the `pendingRecordsAfterError` property is not changed since `isPaused` is true
* let's say the container is resumed right after a poll
* the consumer is resumed
* since the consumer was resumed, `doPoll` returns the next batch of records
* since `isPaused` is false and `pendingRecordsAfterError` is not null, the records returned by the poll are replaced by the records in `pendingRecordsAfterError`
* now we're processing the batch we had before the container was paused, but the consumer position has advanced a batch further, and we have no reference to that batch of records

Makes any sense? Am I missing something?

If that's correct, it seems to me that the problem would be that we're relying on the consumer being paused while we poll and already have records, but the consumer might have been paused and resumed. Maybe we should not resume the consumer if we have pending records?

Reply:
Good catch - I think it should behave as if there are no errors - in other words, don't check `isPaused()`.

Normally, `pause()` does not take effect until the records from the previous poll are all processed.

With my next commit we won't look at `isPaused()` either during the records replacement or in the pause logic itself. i.e. the user call to `pause()` will not take effect until the records from the previous poll are processed, regardless of any errors.

This is a behavior change, but it's a side effect of not doing the seeks on every error.

Reply:
A couple more thoughts on this:

* Are we protected from the same issue in case the consumer resumes automatically after a rebalance?

* I wonder if, in order to simplify things a bit and make it more explicit, we might kill the `isPauseAfterError` flag, and just have different behavior whenever we have `pendingRecordsAfterError != null`. The only part where `pendingRecordsAfterError` is null is between poll and invokeListener, and we don't seem to be using this flag there.

Thanks.

Reply:
>Are we protected from the same issue in case the consumer resumes automatically after a rebalance?

I am not sure what you mean by that; the paused state of any revoked partitions is reset in the `Consumer` so if they are re-assigned, they will not be paused; hence the existing logic to re-pause the entire container if it was paused based on `consumerPaused`. Since we re-pause the TPs within the `onPartitionsAssigned()`, that poll (which is calling the rebalance listener) won't return any records.

Yes, I agree, we can get rid of that boolean.

Reply:
> > Are we protected from the same issue in case the consumer resumes automatically after a rebalance?
> 
> I am not sure what you mean by that; the paused state of any revoked partitions is reset in the `Consumer` so if they are re-assigned, they will not be paused; hence the existing logic to re-pause the entire container if it was paused based on `consumerPaused`. Since we re-pause the TPs within the `onPartitionsAssigned()`, that poll (which is calling the rebalance listener) won't return any records.
> 

Sure, you're right - I forgot we are setting `consumerPaused` to true, so in a rebalance the `RebalanceListener` will pause the consumer again.

I'm just thinking of any situations where the consumer might be resumed in such a way that we might get to `doPoll` with `pendingRecords` before it's paused again. Guess we're good.

I wonder if it might make sense to make an assertion that, if `pendingRecords` is not null and the consumer is not `paused` when we get to the poll, we'd throw an exception just to be on the safe side. But of course, that's mostly my ignorance on the ways the consumer might behave - we can leave that out.

Thanks.

Reply:
I can't think of a scenario where that could happen, but I can add an emergency container stop if the poll returns any records.